In [5]:
# regular imports

import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
import env
from env import host, user, password

# default pandas decimal number display format
pd.options.display.float_format = '{:20,.2f}'.format
pd.set_option('display.max_columns', None)
import warnings
warnings.filterwarnings("ignore")

# Wrangling
from sklearn.model_selection import train_test_split

In [6]:

def get_db_url(db):
    return f'mysql+pymysql://{user}:{password}@{host}/{db}'
    
sql_query = '''
SELECT prop. *,
predictions_2017.logerror,
predictions_2017.transactiondate,
air.airconditioningdesc,
arch.architecturalstyledesc,
build.buildingclassdesc,
heat.heatingorsystemdesc,
land.propertylandusedesc,
story.storydesc,
type.typeconstructiondesc
FROM properties_2017 prop
JOIN (
            SELECT parcelid, MAX(transactiondate) AS max_transactiondate
            FROM predictions_2017
            GROUP BY parcelid) pred USING(parcelid)
JOIN predictions_2017 ON pred.parcelid = predictions_2017.parcelid
AND pred.max_transactiondate = predictions_2017.transactiondate
LEFT JOIN airconditioningtype air USING(airconditioningtypeid)
LEFT JOIN architecturalstyletype arch USING(architecturalstyletypeid)
LEFT JOIN buildingclasstype build USING(buildingclasstypeid)
LEFT JOIN heatingorsystemtype heat USING(heatingorsystemtypeid)
LEFT JOIN propertylandusetype land USING(propertylandusetypeid)
LEFT JOIN storytype story USING(storytypeid)
LEFT JOIN typeconstructiontype type USING(typeconstructiontypeid)
WHERE propertylandusedesc IN ('Single Family Residential' , 'Mobile Home', 'Manufactured, Modular, Prefabricated Homes', 'Patio Home', 'Bungalow', 'Planned Unit Development') 
AND transactiondate <= '2017-12-31'
AND prop.longitude IS NOT NULL
AND prop.latitude IS NOT NULL;
'''
    
def get_zillow_data():
    df = pd.read_sql(sql_query, get_db_url('zillow'))
    df = df.drop(columns='id')
    return df

In [10]:
df = get_zillow_data()

In [11]:
#Summary statistics
df.describe().T

count                  mean  \
parcelid                                54,395.00         12,968,798.63   
airconditioningtypeid                   15,320.00                  2.28   
architecturalstyletypeid                    70.00                  7.10   
basementsqft                                47.00                678.98   
bathroomcnt                             54,395.00                  2.32   
bedroomcnt                              54,395.00                  3.30   
buildingqualitytypeid                   35,647.00                  6.35   
calculatedbathnbr                       54,221.00                  2.33   
decktypeid                                 392.00                 66.00   
finishedfloor1squarefeet                 4,381.00              1,546.23   
calculatedfinishedsquarefeet            54,291.00              1,927.99   
finishedsquarefeet12                    54,085.00              1,930.50   
finishedsquarefeet13                        41.00              1,389.85   
finishedsquarefeet50                     4,381.00              1,566.99   
finishedsquarefeet6                        165.00              1,241.08   
fips                                    54,395.00              6,048.77   
fireplacecnt                             7,234.00                  1.21   
fullbathcnt                             54,221.00                  2.27   
garagecarcnt                            18,027.00                  1.95   
garagetotalsqft                         18,027.00                494.74   
hashottuborspa                           1,514.00                  1.00   
heatingorsystemtypeid                   35,784.00                  3.86   
latitude                                54,395.00         34,027,527.51   
longitude                               54,395.00       -118,199,190.30   
lotsizesquarefeet                       53,956.00             12,575.40   
poolcnt                                 11,447.00                  1.00   
poolsizesum                                865.00                518.31   
pooltypeid10                               443.00                  1.00   
pooltypeid2                              1,071.00                  1.00   
pooltypeid7                             10,357.00                  1.00   
propertylandusetypeid                   54,395.00                261.30   
rawcensustractandblock                  54,395.00         60,490,909.26   
regionidcity                            53,242.00             33,475.75   
regionidcounty                          54,395.00              2,544.69   
regionidneighborhood                    19,743.00            181,066.75   
regionidzip                             54,365.00             96,627.20   
roomcnt                                 54,395.00                  1.83   
storytypeid                                 47.00                  7.00   
threequarterbathnbr                      6,717.00                  1.01   
typeconstructiontypeid                      76.00                  5.97   
unitcnt                                 35,751.00                  1.00   
yardbuildingsqft17                       1,952.00                336.62   
yardbuildingsqft26                          65.00                229.74   
yearbuilt                               54,260.00              1,964.38   
numberofstories                         14,542.00                  1.41   
fireplaceflag                               81.00                  1.00   
structuretaxvaluedollarcnt              54,301.00            198,609.31   
taxvaluedollarcnt                       54,394.00            529,143.53   
assessmentyear                          54,395.00              2,016.00   
landtaxvaluedollarcnt                   54,393.00            330,879.87   
taxamount                               54,391.00              6,460.63   
taxdelinquencyyear                       2,113.00                 14.10   
censustractandblock                     54,265.00 60,498,605,800,781.

In [12]:
#Info on the columns
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 54395 entries, 0 to 54394
Data columns (total 67 columns):
 #   Column                        Non-Null Count  Dtype  
---  ------                        --------------  -----  
 0   parcelid                      54395 non-null  int64  
 1   airconditioningtypeid         15320 non-null  float64
 2   architecturalstyletypeid      70 non-null     float64
 3   basementsqft                  47 non-null     float64
 4   bathroomcnt                   54395 non-null  float64
 5   bedroomcnt                    54395 non-null  float64
 6   buildingclasstypeid           0 non-null      object 
 7   buildingqualitytypeid         35647 non-null  float64
 8   calculatedbathnbr             54221 non-null  float64
 9   decktypeid                    392 non-null    float64
 10  finishedfloor1squarefeet      4381 non-null   float64
 11  calculatedfinishedsquarefeet  54291 non-null  float64
 12  finishedsquarefeet12          54085 non-null  float64
 13  f

In [13]:
print(f'r0ws: {df.shape[0]}')
print(f'co1umns: {df.shape[1]}')

r0ws: 54395
co1umns: 67


In [14]:
for col in df.columns:
    print(col)
    print(df[col].value_counts())

parcelid
14297519    1
12050704    1
11217239    1
13967614    1
12865580    1
           ..
12410258    1
10970834    1
14171698    1
12893364    1
12826780    1
Name: parcelid, Length: 54395, dtype: int64
airconditioningtypeid
 1.00    13578
13.00     1567
 5.00      159
11.00       16
Name: airconditioningtypeid, dtype: int64
architecturalstyletypeid
 7.00    62
 3.00     3
 2.00     2
21.00     2
 8.00     1
Name: architecturalstyletypeid, dtype: int64
basementsqft
  900.00    2
  640.00    2
  100.00    2
  515.00    2
  273.00    2
  912.00    2
  314.00    1
  819.00    1
1,809.00    1
  604.00    1
  126.00    1
  588.00    1
  786.00    1
  669.00    1
  800.00    1
  396.00    1
  112.00    1
  645.00    1
  384.00    1
1,969.00    1
  252.00    1
  600.00    1
1,218.00    1
  280.00    1
  300.00    1
  224.00    1
  512.00    1
  380.00    1
  204.00    1
  200.00    1
   90.00    1
3,112.00    1
  405.00    1
1,252.00    1
  254.00    1
3,560.00    1
  352.00    1
  168.00

Nulls in data

In [15]:
df.isnull().head()

parcelid  airconditioningtypeid  architecturalstyletypeid  basementsqft  \
0     False                   True                      True          True   
1     False                   True                      True          True   
2     False                   True                      True          True   
3     False                   True                      True          True   
4     False                  False                      True          True   

   bathroomcnt  bedroomcnt  buildingclasstypeid  buildingqualitytypeid  \
0        False       False                 True                   True   
1        False       False                 True                   True   
2        False       False                 True                   True   
3        False       False                 True                  False   
4        False       False                 True                  False   

   calculatedbathnbr  decktypeid  finishedfloor1squarefeet  \
0              False        True                      True   
1              False        True                     False   
2              False        True                      True   
3              False        True                      True   
4              False        True                      True   

   calculatedfinishedsquarefeet  finishedsquarefeet12  finishedsquarefeet13  \
0                         False                 False                  True   
1                         False                 False                  True   
2                         False                 False                  True   
3                         False                 False                  True   
4                         False                 False                  True   

   finishedsquarefeet15  finishedsquarefeet50  finishedsquarefeet6   fips  \
0                  True                  True                 True  False   
1                  True                 False                 True  False   
2                  True                  True                 True  False   
3                  True                  True                 True  False   
4                  True                  True                 True  False   

   fireplacecnt  fullbathcnt  garagecarcnt  garagetotalsqft  hashottuborspa  \
0          True        False         False            False            True   
1         False        False         False            False            True   
2          True        False         False            False            True   
3          True        False          True             True            True   
4          True        False          True             True            True   

   heatingorsystemtypeid  latitude  longitude  lotsizesquarefeet  poolcnt  \
0                   True     False      False              False     True   
1                   True     False      False              False     True   
2                   True     False      False              False    False   
3                  False     False      False              False    False   
4                  False     False      False              False    False   

   poolsizesum  pooltypeid10  pooltypeid2  pooltypeid7  \
0         True          True         True         True   
1         True          True         True         True   
2         True          True         True        False   
3         True          True         True        False   
4         True          True         True        False   

   propertycountylandusecode  propertylandusetypeid  propertyzoningdesc  \
0                      False                  False                True   
1                      False                  False                True   
2                      False                  False                True   
3                      False                  False               False   
4                      False                  False               False   

   rawcensustractandblock  regionidcity 

In [16]:
df.head()

parcelid  airconditioningtypeid  architecturalstyletypeid  basementsqft  \
0  14297519                    NaN                       NaN           NaN   
1  17052889                    NaN                       NaN           NaN   
2  14186244                    NaN                       NaN           NaN   
3  12177905                    NaN                       NaN           NaN   
4  12095076                   1.00                       NaN           NaN   

           bathroomcnt           bedroomcnt buildingclasstypeid  \
0                 3.50                 4.00                None   
1                 1.00                 2.00                None   
2                 2.00                 3.00                None   
3                 3.00                 4.00                None   
4                 3.00                 4.00                None   

   buildingqualitytypeid    calculatedbathnbr  decktypeid  \
0                    NaN                 3.50         NaN   
1                    NaN                 1.00         NaN   
2                    NaN                 2.00         NaN   
3                   8.00                 3.00         NaN   
4                   9.00                 3.00         NaN   

   finishedfloor1squarefeet  calculatedfinishedsquarefeet  \
0                       NaN                      3,100.00   
1                  1,465.00                      1,465.00   
2                       NaN                      1,243.00   
3                       NaN                      2,376.00   
4                       NaN                      2,962.00   

   finishedsquarefeet12  finishedsquarefeet13 finishedsquarefeet15  \
0              3,100.00                   NaN                 None   
1              1,465.00                   NaN                 None   
2              1,243.00                   NaN                 None   
3              2,376.00                   NaN                 None   
4              2,962.00                   NaN                 None   

   finishedsquarefeet50  finishedsquarefeet6                 fips  \
0                   NaN                  NaN             6,059.00   
1              1,465.00                  NaN             6,111.00   
2                   NaN                  NaN             6,059.00   
3                   NaN                  NaN             6,037.00   
4                   NaN                  NaN             6,037.00   

          fireplacecnt          fullbathcnt         garagecarcnt  \
0                  NaN                 3.00                 2.00   
1                 1.00                 1.00                 1.00   
2                  NaN                 2.00                 2.00   
3                  NaN                 3.00                  NaN   
4                  NaN                 3.00                  NaN   

       garagetotalsqft  hashottuborspa  heatingorsystemtypeid  \
0               633.00             NaN                    NaN   
1                 0.00             NaN                    NaN   
2               440.00             NaN                    NaN   
3                  NaN             NaN                   2.00   
4                  NaN             NaN                   2.00   

              latitude            longitude    lotsizesquarefeet  \
0        33,634,931.00      -117,869,207.00             4,506.00   
1        34,449,266.00      -119,281,531.00            12,647.00   
2        33,886,168.00      -117,823,170.00             8,432.00   
3        34,245,180.00      -118,240,722.00            13,038.00   
4        34,145,202.00      -118,179,824.00            63,000.00   

               poolcnt  poolsizesum  pooltypeid10  pooltypeid2  \
0                  NaN          NaN           NaN          NaN   
1                  NaN          NaN           NaN          NaN   
2                 1.00          NaN           NaN          NaN   
3                 1.00          NaN           NaN          NaN   
4                 1.00

In [20]:
pd.DataFrame(df.isnull().sum())


0
parcelid                      0
airconditioningtypeid     39075
architecturalstyletypeid  54325
basementsqft              54348
bathroomcnt                   0
...                         ...
buildingclassdesc         54395
heatingorsystemdesc       18611
propertylandusedesc           0
storydesc                 54348
typeconstructiondesc      54319

[67 rows x 1 columns]

Nulls in columns

In [19]:
pd.DataFrame((df.isnull().sum()/df.shape[0]*100)) # Percentage of nulls in each column

0
parcelid                                 0.00
airconditioningtypeid                   71.84
architecturalstyletypeid                99.87
basementsqft                            99.91
bathroomcnt                              0.00
...                                       ...
buildingclassdesc                      100.00
heatingorsystemdesc                     34.21
propertylandusedesc                      0.00
storydesc                               99.91
typeconstructiondesc                    99.86

[67 rows x 1 columns]

In [21]:
nulls_col = pd.DataFrame({'num_rows_missing': df.isnull().sum(), 
              'percent_rows_missing': (df.isnull().sum() / df.shape[0] * 100)})

nulls_col.sort_values(by='num_rows_missing', ascending=False)

num_rows_missing  percent_rows_missing
buildingclasstypeid               54395                100.00
buildingclassdesc                 54395                100.00
finishedsquarefeet15              54395                100.00
finishedsquarefeet13              54354                 99.92
basementsqft                      54348                 99.91
...                                 ...                   ...
latitude                              0                  0.00
fips                                  0                  0.00
bedroomcnt                            0                  0.00
bathroomcnt                           0                  0.00
propertylandusetypeid                 0                  0.00

[67 rows x 2 columns]

In [22]:
def nulls_by_col(df):
    num_missing = df.isnull().sum()
    rows = df.shape[0]
    prcnt_miss = num_missing / rows * 100
    cols_missing = pd.DataFrame({'num_rows_missing': num_missing, 'percent_rows_missing': prcnt_miss})
    return cols_missing.sort_values(by='num_rows_missing', ascending=False)

Nulls by Row

In [23]:
df.isnull().sum(axis=1)

0        36
1        33
2        34
3        32
4        29
         ..
54390    34
54391    33
54392    32
54393    32
54394    34
Length: 54395, dtype: int64

In [24]:
df.head(1)[df.head(1).isna()]

parcelid  airconditioningtypeid  architecturalstyletypeid  basementsqft  \
0       NaN                    NaN                       NaN           NaN   

   bathroomcnt  bedroomcnt buildingclasstypeid  buildingqualitytypeid  \
0          NaN         NaN                None                    NaN   

   calculatedbathnbr  decktypeid  finishedfloor1squarefeet  \
0                NaN         NaN                       NaN   

   calculatedfinishedsquarefeet  finishedsquarefeet12  finishedsquarefeet13  \
0                           NaN                   NaN                   NaN   

  finishedsquarefeet15  finishedsquarefeet50  finishedsquarefeet6  fips  \
0                 None                   NaN                  NaN   NaN   

   fireplacecnt  fullbathcnt  garagecarcnt  garagetotalsqft  hashottuborspa  \
0           NaN          NaN           NaN              NaN             NaN   

   heatingorsystemtypeid  latitude  longitude  lotsizesquarefeet  poolcnt  \
0                    NaN       NaN        NaN                NaN      NaN   

   poolsizesum  pooltypeid10  pooltypeid2  pooltypeid7  \
0          NaN           NaN          NaN          NaN   

  propertycountylandusecode  propertylandusetypeid propertyzoningdesc  \
0                       NaN                    NaN               None   

   rawcensustractandblock  regionidcity  regionidcounty  regionidneighborhood  \
0                     NaN           NaN             NaN                   NaN   

   regionidzip  roomcnt  storytypeid  threequarterbathnbr  \
0          NaN      NaN          NaN                  NaN   

   typeconstructiontypeid  unitcnt  yardbuildingsqft17  yardbuildingsqft26  \
0                     NaN      NaN                 NaN                 NaN   

   yearbuilt  numberofstories  fireplaceflag  structuretaxvaluedollarcnt  \
0        NaN              NaN            NaN                         NaN   

   taxvaluedollarcnt  assessmentyear  landtaxvaluedollarcnt  taxamount  \
0                NaN             NaN                    NaN        NaN   

  taxdelinquencyflag  taxdelinquencyyear  censustractandblock  logerror  \
0               None                 NaN                  NaN       NaN   

  transactiondate airconditioningdesc architecturalstyledesc  \
0             NaN                None                   None   

  buildingclassdesc heatingorsystemdesc propertylandusedesc storydesc  \
0              None                None                 NaN      None   

  typeconstructiondesc  
0                 None

In [25]:
df.isnull().sum(axis=1)


0        36
1        33
2        34
3        32
4        29
         ..
54390    34
54391    33
54392    32
54393    32
54394    34
Length: 54395, dtype: int64

In [26]:
df.isnull().sum(axis=1) / df.shape[1] * 100

0                      53.73
1                      49.25
2                      50.75
3                      47.76
4                      43.28
                ...         
54390                  50.75
54391                  49.25
54392                  47.76
54393                  47.76
54394                  50.75
Length: 54395, dtype: float64

In [27]:
nulls_row = pd.DataFrame({'num_cols_missing': df.isnull().sum(axis=1),
              'percent_cols_missing': df.isnull().sum(axis=1)/df.shape[1]*100})

nulls_row

num_cols_missing  percent_cols_missing
0                    36                 53.73
1                    33                 49.25
2                    34                 50.75
3                    32                 47.76
4                    29                 43.28
...                 ...                   ...
54390                34                 50.75
54391                33                 49.25
54392                32                 47.76
54393                32                 47.76
54394                34                 50.75

[54395 rows x 2 columns]

In [28]:
nulls_row = df.merge(nulls_row,
                        left_index=True,
                        right_index=True)[['parcelid', 'num_cols_missing', 'percent_cols_missing']]

nulls_row

parcelid  num_cols_missing  percent_cols_missing
0      14297519                36                 53.73
1      17052889                33                 49.25
2      14186244                34                 50.75
3      12177905                32                 47.76
4      12095076                29                 43.28
...         ...               ...                   ...
54390  12412492                34                 50.75
54391  11000655                33                 49.25
54392  17239384                32                 47.76
54393  12773139                32                 47.76
54394  12826780                34                 50.75

[54395 rows x 3 columns]

In [29]:
nulls_row.sort_values(by='num_cols_missing', ascending=False)

parcelid  num_cols_missing  percent_cols_missing
39062   14341728                48                 71.64
9287   167686999                48                 71.64
29088  167687839                48                 71.64
11537  167689317                47                 70.15
42974   17188862                47                 70.15
...          ...               ...                   ...
41067   14292517                24                 35.82
17445   14172935                24                 35.82
38906   14193288                24                 35.82
22863   14465914                23                 34.33
40932   14215623                23                 34.33

[54395 rows x 3 columns]

In [30]:
def nulls_by_row(df):
    num_missing = df.isnull().sum(axis=1)
    prcnt_miss = num_missing / df.shape[1] * 100
    rows_missing = pd.DataFrame({'num_cols_missing': num_missing, 'percent_cols_missing': prcnt_miss})
    rows_missing = df.merge(rows_missing,
                        left_index=True,
                        right_index=True)[['parcelid', 'num_cols_missing', 'percent_cols_missing']]
    return rows_missing.sort_values(by='num_cols_missing', ascending=False)

In [32]:
def summarize(df):
    '''
    summarize will take in a single argument (a pandas dataframe) 
    and output to console various statistics on said dataframe, including:
    # .head()
    # .info()
    # .describe()
    # .value_counts()
    # observation of nulls in the dataframe
    '''
    print('SUMMARY REPORT')
    print('=====================================================\n\n')
    print('Dataframe head: ')
    print(df.head(3))
    print('=====================================================\n\n')
    print('Dataframe info: ')
    print(df.info())
    print('=====================================================\n\n')
    print('Dataframe Description: ')
    print(df.describe())
    num_cols = [col for col in df.columns if df[col].dtype != 'O']
    cat_cols = [col for col in df.columns if col not in num_cols]
    print('=====================================================')
    print('DataFrame value counts: ')
    for col in df.columns:
        if col in cat_cols:
            print(df[col].value_counts(), '\n')
        else:
            print(df[col].value_counts(bins=10, sort=False), '\n')
    print('=====================================================')
    print('nulls in dataframe by column: ')
    print(nulls_by_col(df))
    print('=====================================================')
    print('nulls in dataframe by row: ')
    print(nulls_by_row(df))
    print('=====================================================')

In [33]:
summarize(df)

SUMMARY REPORT


Dataframe head: 
   parcelid  airconditioningtypeid  architecturalstyletypeid  basementsqft  \
0  14297519                    NaN                       NaN           NaN   
1  17052889                    NaN                       NaN           NaN   
2  14186244                    NaN                       NaN           NaN   

           bathroomcnt           bedroomcnt buildingclasstypeid  \
0                 3.50                 4.00                None   
1                 1.00                 2.00                None   
2                 2.00                 3.00                None   

   buildingqualitytypeid    calculatedbathnbr  decktypeid  \
0                    NaN                 3.50         NaN   
1                    NaN                 1.00         NaN   
2                    NaN                 2.00         NaN   

   finishedfloor1squarefeet  calculatedfinishedsquarefeet  \
0                       NaN                      3,100.00   
1                

Our accomplishments so far

We have:

- Acquired our data
- Examined the structure and integrity of the data
- We have observed univariate descriptive statistics
- We have examined null values (in total, by column, and by row)
- We have created functions for these processes


What we still need to do:

- Determine what columns can be saved by filling in nulls.
- Create functions that identify upper bound outliers and lower bound outliers
- Create functions that act on our outliers (drop them, compress them, etc.)
- Create functions/processes that act on nulls beyond simple removal


Fill in null values.

In [46]:
df["fireplacecnt"].fillna( 0 , inplace = True)

In [54]:
df['fireplaceflag'] = np.where(df.fireplacecnt > 0, 1 , 0)

In [55]:
df.fireplaceflag.value_counts()

0    47161
1     7234
Name: fireplaceflag, dtype: int64

In [48]:
df.fireplacecnt.value_counts()

0.00    47161
1.00     6025
2.00      942
3.00      229
4.00       35
5.00        3
Name: fireplacecnt, dtype: int64

In [60]:
df.loc[df['fireplacecnt'] > 0]

parcelid  airconditioningtypeid  architecturalstyletypeid  \
1      17052889                    NaN                       NaN   
8      13944538                  13.00                       NaN   
12     14387959                    NaN                       NaN   
16     14314879                    NaN                       NaN   
24     14344863                    NaN                       NaN   
...         ...                    ...                       ...   
54025  14425750                    NaN                       NaN   
54046  14389748                   1.00                       NaN   
54053  14387399                  13.00                       NaN   
54093  13976300                    NaN                       NaN   
54392  17239384                    NaN                       NaN   

       basementsqft          bathroomcnt           bedroomcnt  \
1               NaN                 1.00                 2.00   
8               NaN                 2.50                 3.00   
12              NaN                 2.50                 3.00   
16              NaN                 4.50                 6.00   
24              NaN                 2.50                 4.00   
...             ...                  ...                  ...   
54025           NaN                 2.00                 4.00   
54046           NaN                 2.50                 3.00   
54053           NaN                 2.00                 3.00   
54093           NaN                 2.00                 3.00   
54392           NaN                 2.00                 4.00   

      buildingclasstypeid  buildingqualitytypeid    calculatedbathnbr  \
1                    None                    NaN                 1.00   
8                    None                    NaN                 2.50   
12                   None                    NaN                 2.50   
16                   None                    NaN                 4.50   
24                   None                    NaN                 2.50   
...                   ...                    ...                  ...   
54025                None                    NaN                 2.00   
54046                None                    NaN                 2.50   
54053                None                    NaN                 2.00   
54093                None                    NaN                 2.00   
54392                None                    NaN                 2.00   

       decktypeid  finishedfloor1squarefeet  calculatedfinishedsquarefeet  \
1             NaN                  1,465.00                      1,465.00   
8             NaN                       NaN                      1,340.00   
12            NaN                       NaN                      2,074.00   
16            NaN                       NaN                      3,870.00   
24            NaN                       NaN                      1,979.00   
...           ...                       ...                           ...   
54025         NaN                       NaN                      1,358.00   
54046         NaN                       NaN                      2,026.00   
54053         NaN                       NaN                      1,789.00   
54093         NaN                       NaN                      1,156.00   
54392         NaN                  1,612.00                      1,612.00   

       finishedsquarefeet12  finishedsquarefeet13 finishedsquarefeet15  \
1                  1,465.00                   NaN                 None   
8                  1,340.00                   NaN                 None   
12                 2,074.00                   NaN                 None   
16                 3,870.00                   NaN                 None   
24                 1,979.00                   NaN                 None   
...                     ...                   ...                  ...   
54025              1,358.00                   NaN                 None   
54046          

In [37]:
# Data Dictionary of zillow data. Will help determine what to save
pd.set_option('display.max_colwidth', None)
raw_data_dictionary = pd.read_clipboard()
raw_data_dictionary

Feature  \
0          'airconditioningtypeid'   
1       'architecturalstyletypeid'   
2                   'basementsqft'   
3                    'bathroomcnt'   
4                     'bedroomcnt'   
5          'buildingqualitytypeid'   
6            'buildingclasstypeid'   
7              'calculatedbathnbr'   
8                     'decktypeid'   
9            'threequarterbathnbr'   
10      'finishedfloor1squarefeet'   
11  'calculatedfinishedsquarefeet'   
12           'finishedsquarefeet6'   
13          'finishedsquarefeet12'   
14          'finishedsquarefeet13'   
15          'finishedsquarefeet15'   
16          'finishedsquarefeet50'   
17                          'fips'   
18                  'fireplacecnt'   
19                 'fireplaceflag'   
20                   'fullbathcnt'   
21                  'garagecarcnt'   
22               'garagetotalsqft'   
23                'hashottuborspa'   
24         'heatingorsystemtypeid'   
25                      'latitude'   
26                     'longitude'   
27             'lotsizesquarefeet'   
28               'numberofstories'   
29                      'parcelid'   
30                       'poolcnt'   
31                   'poolsizesum'   
32                  'pooltypeid10'   
33                   'pooltypeid2'   
34                   'pooltypeid7'   
35     'propertycountylandusecode'   
36         'propertylandusetypeid'   
37            'propertyzoningdesc'   
38        'rawcensustractandblock'   
39           'censustractandblock'   
40                'regionidcounty'   
41                  'regionidcity'   
42                   'regionidzip'   
43          'regionidneighborhood'   
44                       'roomcnt'   
45                   'storytypeid'   
46        'typeconstructiontypeid'   
47                       'unitcnt'   
48            'yardbuildingsqft17'   
49            'yardbuildingsqft26'   
50                     'yearbuilt'   
51             'taxvaluedollarcnt'   
52    'structuretaxvaluedollarcnt'   
53         'landtaxvaluedollarcnt'   
54                     'taxamount'   
55                'assessmentyear'   
56            'taxdelinquencyflag'   
57            'taxdelinquencyyear'   

                                                                                                               Description  
0                                                                      Type of cooling system present in the home (if any)  
1                                                Architectural style of the home (i.e. ranch, colonial, split-level, etc…)  
2                                                               Finished living area below or partially below ground level  
3                                                               Number of bathrooms in home including fractional bathrooms  
4                                                                                              Number of bedrooms in home   
5                                    Overall assessment of condition of the building from best (lowest) to worst (highest)  
6                                                     The building framing type (steel frame, wood frame, concrete/brick)   
7                                                                Number of bathrooms in home including fractional bathroom  
8                                                                                  Type of deck (if any) present on parcel  
9                                                                Number of 3/4 bathrooms in house (shower + sink + toilet)  
10                                                 Size of the finished living area on the first (entry) floor of the home  
11                                                                      Calculated total finished living area of the home   
12                                                                                       Base unfinished and finished area  
13                                

Takeaways

- Fill in fireplaceflag with values 0 or 1. Filled in values does not match 'fireplacecnt' total.
- Fill in nulls with 0 in columns: 'basementsqft', 'fireplacecnt', 

- Drop 'hashottuborspa' due to column is the sum of columns 'pooltypeid2' and 'pooltypeid10'
- Drop 'poolcnt' due to sum (11447) does not match totalof columns of 'pooltypeid2' and 'pooltypeid7' (11374) short 73 pools, create new column 'haspool' with values 0 meaning No or 1 meaning yes.

In [61]:
df.poolcnt.value_counts()

1.00    11447
Name: poolcnt, dtype: int64

In [62]:
df.pooltypeid10.value_counts()

1.00    443
Name: pooltypeid10, dtype: int64

In [63]:
df.pooltypeid2.value_counts()

1.00    1071
Name: pooltypeid2, dtype: int64

In [64]:
df.pooltypeid7.value_counts()

1.00    10357
Name: pooltypeid7, dtype: int64

In [65]:
df.hashottuborspa.value_counts()

1.00    1514
Name: hashottuborspa, dtype: int64

In [66]:
df.loc[df['pooltypeid2'] > 0]

parcelid  airconditioningtypeid  architecturalstyletypeid  \
46     14434171                    NaN                       NaN   
147    14350624                    NaN                       NaN   
171    14394667                    NaN                       NaN   
189    14391477                    NaN                       NaN   
291    14081668                    NaN                       NaN   
...         ...                    ...                       ...   
53671  14461441                    NaN                       NaN   
53816  14340406                    NaN                       NaN   
54009  14434009                    NaN                       NaN   
54029  14436536                    NaN                       NaN   
54093  13976300                    NaN                       NaN   

       basementsqft          bathroomcnt           bedroomcnt  \
46              NaN                 6.50                 5.00   
147             NaN                 4.00                 6.00   
171             NaN                 3.00                 4.00   
189             NaN                 2.50                 4.00   
291             NaN                 4.00                 4.00   
...             ...                  ...                  ...   
53671           NaN                 6.50                 6.00   
53816           NaN                 4.50                 5.00   
54009           NaN                 4.00                 3.00   
54029           NaN                 4.00                 5.00   
54093           NaN                 2.00                 3.00   

      buildingclasstypeid  buildingqualitytypeid    calculatedbathnbr  \
46                   None                    NaN                 6.50   
147                  None                    NaN                 4.00   
171                  None                    NaN                 3.00   
189                  None                    NaN                 2.50   
291                  None                    NaN                 4.00   
...                   ...                    ...                  ...   
53671                None                    NaN                 6.50   
53816                None                    NaN                 4.50   
54009                None                    NaN                 4.00   
54029                None                    NaN                 4.00   
54093                None                    NaN                 2.00   

       decktypeid  finishedfloor1squarefeet  calculatedfinishedsquarefeet  \
46            NaN                       NaN                      9,586.00   
147           NaN                       NaN                      3,733.00   
171           NaN                       NaN                      3,309.00   
189           NaN                       NaN                      2,357.00   
291           NaN                       NaN                      3,186.00   
...           ...                       ...                           ...   
53671         NaN                       NaN                      4,767.00   
53816         NaN                       NaN                      4,038.00   
54009         NaN                       NaN                      3,454.00   
54029         NaN                       NaN                      3,542.00   
54093         NaN                       NaN                      1,156.00   

       finishedsquarefeet12  finishedsquarefeet13 finishedsquarefeet15  \
46                 9,586.00                   NaN                 None   
147                3,733.00                   NaN                 None   
171                3,309.00                   NaN                 None   
189                2,357.00                   NaN                 None   
291                3,186.00                   NaN                 None   
...                     ...                   ...                  ...   
53671              4,767.00                   NaN                 None   
53816          

In [67]:
df.loc[df['pooltypeid7'] > 0]

parcelid  airconditioningtypeid  architecturalstyletypeid  \
2      14186244                    NaN                       NaN   
3      12177905                    NaN                       NaN   
4      12095076                   1.00                       NaN   
7      11104527                   1.00                       NaN   
16     14314879                    NaN                       NaN   
...         ...                    ...                       ...   
54369  10867042                    NaN                       NaN   
54370  13089085                   1.00                       NaN   
54378  12401161                    NaN                       NaN   
54379  11068972                   1.00                       NaN   
54381  12669704                    NaN                       NaN   

       basementsqft          bathroomcnt           bedroomcnt  \
2               NaN                 2.00                 3.00   
3               NaN                 3.00                 4.00   
4               NaN                 3.00                 4.00   
7               NaN                 3.00                 4.00   
16              NaN                 4.50                 6.00   
...             ...                  ...                  ...   
54369           NaN                 2.00                 3.00   
54370           NaN                 3.00                 4.00   
54378           NaN                 3.00                 3.00   
54379           NaN                 5.00                 4.00   
54381           NaN                 2.00                 4.00   

      buildingclasstypeid  buildingqualitytypeid    calculatedbathnbr  \
2                    None                    NaN                 2.00   
3                    None                   8.00                 3.00   
4                    None                   9.00                 3.00   
7                    None                   8.00                 3.00   
16                   None                    NaN                 4.50   
...                   ...                    ...                  ...   
54369                None                   7.00                 2.00   
54370                None                   8.00                 3.00   
54378                None                   9.00                 3.00   
54379                None                  11.00                 5.00   
54381                None                   8.00                 2.00   

       decktypeid  finishedfloor1squarefeet  calculatedfinishedsquarefeet  \
2             NaN                       NaN                      1,243.00   
3             NaN                       NaN                      2,376.00   
4             NaN                       NaN                      2,962.00   
7             NaN                       NaN                      2,540.00   
16            NaN                       NaN                      3,870.00   
...           ...                       ...                           ...   
54369         NaN                       NaN                      2,141.00   
54370         NaN                       NaN                      1,859.00   
54378         NaN                       NaN                      3,539.00   
54379         NaN                       NaN                      5,210.00   
54381         NaN                       NaN                      2,166.00   

       finishedsquarefeet12  finishedsquarefeet13 finishedsquarefeet15  \
2                  1,243.00                   NaN                 None   
3                  2,376.00                   NaN                 None   
4                  2,962.00                   NaN                 None   
7                  2,540.00                   NaN                 None   
16                 3,870.00                   NaN                 None   
...                     ...                   ...                  ...   
54369              2,141.00                   NaN                 None   
54370          

In [70]:
df[(df.pooltypeid7 > 0) | (df.pooltypeid2 > 0)]

parcelid  airconditioningtypeid  architecturalstyletypeid  \
2      14186244                    NaN                       NaN   
3      12177905                    NaN                       NaN   
4      12095076                   1.00                       NaN   
7      11104527                   1.00                       NaN   
16     14314879                    NaN                       NaN   
...         ...                    ...                       ...   
54369  10867042                    NaN                       NaN   
54370  13089085                   1.00                       NaN   
54378  12401161                    NaN                       NaN   
54379  11068972                   1.00                       NaN   
54381  12669704                    NaN                       NaN   

       basementsqft          bathroomcnt           bedroomcnt  \
2               NaN                 2.00                 3.00   
3               NaN                 3.00                 4.00   
4               NaN                 3.00                 4.00   
7               NaN                 3.00                 4.00   
16              NaN                 4.50                 6.00   
...             ...                  ...                  ...   
54369           NaN                 2.00                 3.00   
54370           NaN                 3.00                 4.00   
54378           NaN                 3.00                 3.00   
54379           NaN                 5.00                 4.00   
54381           NaN                 2.00                 4.00   

      buildingclasstypeid  buildingqualitytypeid    calculatedbathnbr  \
2                    None                    NaN                 2.00   
3                    None                   8.00                 3.00   
4                    None                   9.00                 3.00   
7                    None                   8.00                 3.00   
16                   None                    NaN                 4.50   
...                   ...                    ...                  ...   
54369                None                   7.00                 2.00   
54370                None                   8.00                 3.00   
54378                None                   9.00                 3.00   
54379                None                  11.00                 5.00   
54381                None                   8.00                 2.00   

       decktypeid  finishedfloor1squarefeet  calculatedfinishedsquarefeet  \
2             NaN                       NaN                      1,243.00   
3             NaN                       NaN                      2,376.00   
4             NaN                       NaN                      2,962.00   
7             NaN                       NaN                      2,540.00   
16            NaN                       NaN                      3,870.00   
...           ...                       ...                           ...   
54369         NaN                       NaN                      2,141.00   
54370         NaN                       NaN                      1,859.00   
54378         NaN                       NaN                      3,539.00   
54379         NaN                       NaN                      5,210.00   
54381         NaN                       NaN                      2,166.00   

       finishedsquarefeet12  finishedsquarefeet13 finishedsquarefeet15  \
2                  1,243.00                   NaN                 None   
3                  2,376.00                   NaN                 None   
4                  2,962.00                   NaN                 None   
7                  2,540.00                   NaN                 None   
16                 3,870.00                   NaN                 None   
...                     ...                   ...                  ...   
54369              2,141.00                   NaN                 None   
54370          

In [75]:
def conditions(df):
    if (df['pooltypeid7'] > 0) or (df['pooltypeid2'] > 0) :
        return 1
    else:
        return 0

In [77]:
df['haspool'] = df.apply(conditions, axis=1)

In [78]:
df.haspool.value_counts()

0    42967
1    11428
Name: haspool, dtype: int64

In [81]:
df.head()

parcelid  airconditioningtypeid  architecturalstyletypeid  basementsqft  \
0  14297519                    NaN                       NaN           NaN   
1  17052889                    NaN                       NaN           NaN   
2  14186244                    NaN                       NaN           NaN   
3  12177905                    NaN                       NaN           NaN   
4  12095076                   1.00                       NaN           NaN   

           bathroomcnt           bedroomcnt buildingclasstypeid  \
0                 3.50                 4.00                None   
1                 1.00                 2.00                None   
2                 2.00                 3.00                None   
3                 3.00                 4.00                None   
4                 3.00                 4.00                None   

   buildingqualitytypeid    calculatedbathnbr  decktypeid  \
0                    NaN                 3.50         NaN   
1                    NaN                 1.00         NaN   
2                    NaN                 2.00         NaN   
3                   8.00                 3.00         NaN   
4                   9.00                 3.00         NaN   

   finishedfloor1squarefeet  calculatedfinishedsquarefeet  \
0                       NaN                      3,100.00   
1                  1,465.00                      1,465.00   
2                       NaN                      1,243.00   
3                       NaN                      2,376.00   
4                       NaN                      2,962.00   

   finishedsquarefeet12  finishedsquarefeet13 finishedsquarefeet15  \
0              3,100.00                   NaN                 None   
1              1,465.00                   NaN                 None   
2              1,243.00                   NaN                 None   
3              2,376.00                   NaN                 None   
4              2,962.00                   NaN                 None   

   finishedsquarefeet50  finishedsquarefeet6                 fips  \
0                   NaN                  NaN             6,059.00   
1              1,465.00                  NaN             6,111.00   
2                   NaN                  NaN             6,059.00   
3                   NaN                  NaN             6,037.00   
4                   NaN                  NaN             6,037.00   

          fireplacecnt          fullbathcnt         garagecarcnt  \
0                 0.00                 3.00                 2.00   
1                 1.00                 1.00                 1.00   
2                 0.00                 2.00                 2.00   
3                 0.00                 3.00                  NaN   
4                 0.00                 3.00                  NaN   

       garagetotalsqft  hashottuborspa  heatingorsystemtypeid  \
0               633.00             NaN                    NaN   
1                 0.00             NaN                    NaN   
2               440.00             NaN                    NaN   
3                  NaN             NaN                   2.00   
4                  NaN             NaN                   2.00   

              latitude            longitude    lotsizesquarefeet  \
0        33,634,931.00      -117,869,207.00             4,506.00   
1        34,449,266.00      -119,281,531.00            12,647.00   
2        33,886,168.00      -117,823,170.00             8,432.00   
3        34,245,180.00      -118,240,722.00            13,038.00   
4        34,145,202.00      -118,179,824.00            63,000.00   

               poolcnt  poolsizesum  pooltypeid10  pooltypeid2  \
0                  NaN          NaN           NaN          NaN   
1                  NaN          NaN           NaN          NaN   
2                 1.00          NaN           NaN          NaN   
3                 1.00          NaN           NaN          NaN   
4                 1.00

In [85]:
df.pooltypeid10.value_counts()

1.00    443
Name: pooltypeid10, dtype: int64

In [ ]:
df["hashottuborspa"].fillna( 0 , inplace = True)
df["pooltypeid10"].fillna( 0 , inplace = True)
df["pooltypeid7"].fillna( 0 , inplace = True)
df["pooltypeid10"].fillna( 0 , inplace = True)
df["pooltypeid2"].fillna( 0 , inplace = True)
df["pooltypeid10"].fillna( 0 , inplace = True)

